### PR TITLE
 Adding linux support for perf tests  

### PR DESCRIPTION
**Pull request by dsgouda**
•Modifying python commands to run for Linux runs
•Modifying xunit perf runner to set up execution environment for running perf tests
•Modifying test runtime dependencies to restore the cli version of xunit perf runner for Linux
•Installing dotnet cli per correlation id (by referring to version number specified under xunit performance repo) 

P.S.: xunit perffornance package version 0032 is yet to be published
@lt72 @MattGal

### CODE REVIEW

File: src/Microsoft.DotNet.Build.CloudTestTasks/PackageFiles/CloudTest.targets [hunk 1]
```diff
@@ -203,8 +203,11 @@
   </Target>
 
   <Target Name="CreatePerfTestListJson" DependsOnTargets="CreateAzureStorage" Condition="'$(Performance)' == 'true'">
-    <PropertyGroup>
-      <OtherRunnerScriptArgs>--perf-runner Microsoft.DotNet.xunit.performance.runner.Windows $(OtherRunnerScriptArgs)</OtherRunnerScriptArgs>
+    <PropertyGroup Condition="'$(TargetsWindows)' == 'true'">
+      <OtherRunnerScriptArgs>--perf-runner Microsoft.DotNet.xunit.performance.runner.Windows --osgroup windows $(OtherRunnerScriptArgs)</OtherRunnerScriptArgs>
+    </PropertyGroup>
```
Comment (MattGal) [inline src/Microsoft.DotNet.Build.CloudTestTasks/PackageFiles/CloudTest.targets]: Since these properties are identical except for OSGroup, and there's literally a variable you could toLower called OSGroup, why not make these work?  (OSGroup would be Windows_NT).  Failing that, could still refactor this to using this logic to set $(OSGroupName) and set OtherRunnerScriptArgs once.

**Comment by dsgouda** [comment]
@MattGal PTAL at addressed comments in new commit

**Comment by dsgouda** [comment]
cc: @ChadNedzlek @karajas

File: src/Microsoft.DotNet.Build.CloudTestTasks/PackageFiles/CloudTest.targets [hunk 2]
```diff
@@ -108,7 +108,7 @@
   <Target Name="PreCloudBuild">
     <!-- copy runner scripts so they can be uploaded as supplemental payload -->
     <ItemGroup>
-      <RunnerScripts Include="$(ToolsDir)RunnerScripts/**/*.py" />
+      <RunnerScripts Include="$(ToolsDir)RunnerScripts/**/*.*" />
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/PackageFiles/CloudTest.targets]: Don't just include everything. There will be random pre-compiled things in this folder, and all sorts of temp files.  Even the *.py is pretty dangerous, but anything other than that should be included specifically.

File: src/Microsoft.DotNet.Build.CloudTestTasks/RunnerScripts/xunitrunner-perf/ubuntu-dotnet-local-install.sh [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+echo $DESTDIR
+if [[ -z $DESTDIR ]]
+then
+     usage
+     exit 1
+fi
+
+echo "Initiating local dotnet install"
+wget https://raw.githubusercontent.com/dotnet/cli/rel/1.0.0/scripts/obtain/dotnet-install.sh
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/RunnerScripts/xunitrunner-perf/ubuntu-dotnet-local-install.sh]: If my understanding of this line is correct, it's fetching a hot copy of this script, which might cause runs to start failing unexpectedly, even running the same thing a second time might produce different results (if there is a new latest change in this branch).  Picking a particular version will lead to more stable and reliable tests.

File: src/Microsoft.DotNet.Build.CloudTestTasks/RunnerScripts/xunitrunner-perf/xunitrunner-perf.py [hunk 4]
```diff
@@ -74,9 +74,41 @@ def _prepare_execution_environment(settings, framework_in_tpa, assembly_list_nam
...
+        s = os.path.join(src, item)
+        d = os.path.join(dest, item)
+        if os.path.isfile(s):
+            shutil.copy2(s, d)
+
+def _prepare_linux_env_for_perf(correlation_dir, xunit_perf_drop, test_location):
+    # copy over the cli runner and its dependencies for execution
+    _copy_files_to_dest(xunit_perf_drop, test_location)
+    xunit_perf_deps = os.path.join(correlation_dir,"Microsoft.DotNet.xunit.performance.run.core")
+    xunit_perf_deps = os.path.join(xunit_perf_deps, os.listdir(xunit_perf_deps)[0])
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/RunnerScripts/xunitrunner-perf/xunitrunner-perf.py]: I'd probably have some sort of log or warning to make sure this directory only has a single file in it.

File: src/Microsoft.DotNet.Build.CloudTestTasks/RunnerScripts/xunitrunner-perf/xunitrunner-perf.py [hunk 5]
```diff
@@ -74,9 +74,41 @@ def _prepare_execution_environment(settings, framework_in_tpa, assembly_list_nam
...
+    xunit_perf_deps = os.path.join(*[xunit_perf_deps, "lib", "dotnet"])
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/RunnerScripts/xunitrunner-perf/xunitrunner-perf.py]: The *[] thing is weird. You should just be able to pass the three arguments.

File: src/Microsoft.DotNet.Build.CloudTestTasks/PackageFiles/CloudTest.targets (hunk 2 above)
Comment (dsgouda) [inline src/Microsoft.DotNet.Build.CloudTestTasks/PackageFiles/CloudTest.targets]: As of now I believe there are roughly 4 files and I am adding another, subject to more number of files based on any specific operations we would like to perform based on perf, func, stress, etc. Adding each file by name therefore sounds a little risky. Also, I am modifying *.py to ensure the .sh file is also included. Is there a cleaner way to accomplish this.

File: src/Microsoft.DotNet.Build.CloudTestTasks/RunnerScripts/xunitrunner-perf/xunitrunner-perf.py (hunk 5 above)
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/RunnerScripts/xunitrunner-perf/xunitrunner-perf.py]: I feel like this would be easier to read if it was one (large) join command broken up by line. It's hard to follow which pieces are being glommed on where, and what the final string will look like.

File: src/Microsoft.DotNet.Build.CloudTestTasks/RunnerScripts/xunitrunner-perf/ubuntu-dotnet-local-install.sh (hunk 3 above)
Comment (dsgouda) [inline src/Microsoft.DotNet.Build.CloudTestTasks/RunnerScripts/xunitrunner-perf/ubuntu-dotnet-local-install.sh]: To the dotnet-install script I am explicitly passing the version number which in turn is set in our xunit-performance repo in the DotNetCliVersion.txt file, I believe that should ensure control over versions of dotnet cli that will be installed.

File: src/Microsoft.DotNet.Build.CloudTestTasks/PackageFiles/CloudTest.targets (hunk 2 above)
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/PackageFiles/CloudTest.targets]: *.* means that whatever is in these directories, even if it's not checked into git, will end up deployed. That means deploying from _your_ machine will produce different results than deploying from _my_ machine, which is bad.  Each individual file should be included in here, so that the build/deployment is predictable and repeatable.

And each of the 4 files (and realistically, I'd like to see the *.py exploded as well, but it's far less likely a rogue .py file will be here than, say, some random *.exe that is auto-ignored by git)

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/test-runtime/project.json [hunk 6]
```diff
@@ -11,12 +11,14 @@
...
       "dependencies": {
+        "Microsoft.DotNet.xunit.performance.analysis.cli": "1.0.0-alpha-build0032",
+        "Microsoft.DotNet.xunit.performance.runner.cli": "1.0.0-alpha-build0032",
         "Microsoft.NETCore.Platforms": "1.0.1-rc2-24027",
         "Microsoft.NETCore.TestHost": "1.0.0-rc2-24027",
         "Microsoft.NETCore.Console": "1.0.0-rc2-24027",
         "xunit.console.netcore": "1.0.2-prerelease-00101"
       },
-      "imports": "portable-net45+win8"
+      "imports": ["portable-net45+win8", "netstandard1.3"]
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/test-runtime/project.json]: I think this is a pretty big change (adding a whole new framework dependency). I'm not sure what the ramifications are.

File: src/Microsoft.DotNet.Build.CloudTestTasks/RunnerScripts/xunitrunner-perf/xunitrunner-perf.py [hunk 7]
```diff
@@ -244,19 +288,21 @@ def _run_xunit_from_execution(settings, test_dll, xunit_test_type, args):
...
-        _write_output_path(etl_zip, settings)
+        if xunit_test_type == xunit.XUNIT_CONFIG_PERF_WINDOWS:
+            # only windows runs would generate the etl file
+            # archive the ETL file and upload it
+            etl_file = os.path.join(test_location, "latest-perf-build.etl")
+            etl_zip = os.path.join(test_location, "latest-perf-build.zip")
+            log.info("Compressing {} into {}".format(etl_file, etl_zip))
+            zip_script.zipFilesAndFolders(etl_zip, [etl_file], True, True)
+            log.info("Uploading ETL from {}".format(etl_zip))
+            _write_output_path(etl_zip, settings)
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/RunnerScripts/xunitrunner-perf/xunitrunner-perf.py]: It might be useful to log what this function returns (so that we know where we put the file), in case we can't find it later for some reason.

File: src/Microsoft.DotNet.Build.CloudTestTasks/RunnerScripts/xunitrunner-perf/xunitrunner-perf.py [hunk 8]
```diff
@@ -144,12 +179,20 @@ def _copy_package_files(assembly_list, build_drop, test_location, coreroot_locat
...
     payload_dir = fix_path(os.getenv('HELIX_CORRELATION_PAYLOAD'))
-    perf_analysis_version = (next(os.walk(os.path.join(payload_dir, 'Microsoft.DotNet.xunit.performance.analysis')))[1])[0]
-    xmlconvertorpath = os.path.join(*[payload_dir, 'Microsoft.DotNet.xunit.performance.analysis', perf_analysis_version, 'tools', 'xunit.performance.analysis.exe'])
+    xmlconvertorpath = ''
+
+    if xunit_test_type == xunit.XUNIT_CONFIG_PERF_WINDOWS:
+        perf_analysis_version = (next(os.walk(os.path.join(payload_dir, 'Microsoft.DotNet.xunit.performance.analysis')))[1])[0]
+        xmlconvertorpath = os.path.join(*[payload_dir, 'Microsoft.DotNet.xunit.performance.analysis', perf_analysis_version, 'tools', 'xunit.performance.analysis.exe'])
+    elif xunit_test_type == xunit.XUNIT_CONFIG_PERF_LINUX:
+        perf_analysis_version = (next(os.walk(os.path.join(payload_dir, 'Microsoft.DotNet.xunit.performance.analysis.cli')))[1])[0]
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/RunnerScripts/xunitrunner-perf/xunitrunner-perf.py]: This line is quite hard to follow in my head... What is it actually returning?  is there only a single file/forlder there (since otherwise, walk is going to return the files in whatever order it wants).  What are the [1] and [0] returning (it's hard to tell with the 4 parenthesis all squished together at the end there.  This should probably be broken up into 3-4 lines with good variable names (or put into a function) to make it clear what's actually going on here.

File: src/Microsoft.DotNet.Build.CloudTestTasks/RunnerScripts/xunitrunner-perf/xunitrunner-perf.py [hunk 9]
```diff
@@ -86,16 +118,19 @@ def _prepare_perf_execution_environment(settings, perf_runner):
...
-    shutil.copy2(os.path.join(xunit_perf_drop, "xunit.performance.logger.exe"), test_location)
-    shutil.copy2(os.path.join(xunit_perf_drop, "xunit.runner.utility.desktop.dll"), test_location)
-    shutil.copy2(os.path.join(xunit_perf_drop, "ProcDomain.dll"), test_location)
-    shutil.copy2(os.path.join(xunit_perf_drop, "Microsoft.Diagnostics.Tracing.TraceEvent.dll"), test_location)
+
+    if osgroup.lower().startswith('windows'):
+        _prepare_windows_env_for_perf(xunit_perf_drop, test_location)
+    elif osgroup.lower().startswith('linux'):
+        _prepare_linux_env_for_perf(correlation_dir, xunit_perf_drop, test_location)
     # copy the architecture specific subdirectories
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/RunnerScripts/xunitrunner-perf/xunitrunner-perf.py]: And "else, log.error" is probably in order in case something unexpected is passed happens.

File: src/Microsoft.DotNet.Build.CloudTestTasks/RunnerScripts/xunitrunner-perf/xunitrunner-perf.py [hunk 10]
```diff
@@ -144,12 +179,20 @@ def _copy_package_files(assembly_list, build_drop, test_location, coreroot_locat
...
+
+    if xunit_test_type == xunit.XUNIT_CONFIG_PERF_WINDOWS:
+        perf_analysis_version = (next(os.walk(os.path.join(payload_dir, 'Microsoft.DotNet.xunit.performance.analysis')))[1])[0]
+        xmlconvertorpath = os.path.join(*[payload_dir, 'Microsoft.DotNet.xunit.performance.analysis', perf_analysis_version, 'tools', 'xunit.performance.analysis.exe'])
+    elif xunit_test_type == xunit.XUNIT_CONFIG_PERF_LINUX:
+        perf_analysis_version = (next(os.walk(os.path.join(payload_dir, 'Microsoft.DotNet.xunit.performance.analysis.cli')))[1])[0]
+        dotnet_cli_exec = os.path.join(*(settings.correlation_payload_dir, "dotnet_cli", "dotnet"))
+        xmlconvertorpath = dotnet_cli_exec +' '+ os.path.join(*[payload_dir, 'Microsoft.DotNet.xunit.performance.analysis.cli', perf_analysis_version, 'lib', 'netstandard1.3', 'Microsoft.DotNet.xunit.performance.analysis.cli.dll'])
+
     xmlCmd = xmlconvertorpath+' -csv '+os.path.join(workitem_dir, 'results.csv')+' '+results_location
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/RunnerScripts/xunitrunner-perf/xunitrunner-perf.py]: Another "else error" is called for, in case it's an unexpected value (like XUNIT_CONFIG_PERF from an old repro or something)

File: src/Microsoft.DotNet.Build.CloudTestTasks/RunnerScripts/xunitrunner-perf/ubuntu-dotnet-local-install.sh [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+chmod 777 dotnet-install.sh
+wget https://raw.githubusercontent.com/Microsoft/xunit-performance/master/DotNetCliVersion.txt
+versionnum=$(cat DotNetCliVersion.txt)
+mkdir $DESTDIR
+source dotnet-install.sh -i $DESTDIR -v $versionnum
```
Comment (karajas) [inline src/Microsoft.DotNet.Build.CloudTestTasks/RunnerScripts/xunitrunner-perf/ubuntu-dotnet-local-install.sh]: I know this is an isolated case but what if a conflicting version of the cli exists on the end machine?

Comment (karajas) [inline src/Microsoft.DotNet.Build.CloudTestTasks/RunnerScripts/xunitrunner-perf/ubuntu-dotnet-local-install.sh]: We should maybe include an uninstall step in this process.

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/test-runtime/project.json (hunk 6 above)
Comment (karajas) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/test-runtime/project.json]: Thing to note here is this will affect all test builds in repos that take a dependency on Build.Tasks ( I believe this is from https://github.com/dotnet/buildtools/pull/488/files )

File: src/Microsoft.DotNet.Build.CloudTestTasks/RunnerScripts/xunitrunner-perf/ubuntu-dotnet-local-install.sh (hunk 11 above)
Comment (dsgouda) [inline src/Microsoft.DotNet.Build.CloudTestTasks/RunnerScripts/xunitrunner-perf/ubuntu-dotnet-local-install.sh]: The installation will be put the payload directory of that particular correlation id and should not affect the overall state of the machine, in other words this is a local or temporary installation which means when the correlation payload is cleared out the installation should automatically deleted

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/test-runtime/project.json (hunk 6 above)
Comment (dsgouda) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/test-runtime/project.json]: Per conversation with @joperezr this should not pose a problem and given the constraints of restoring the required packages this is an acceptable solution

Comment (joperezr) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/test-runtime/project.json]: AFAIK this will not add a new framework dependency, but instead just use `netstandard1.3` in the case where `dnxcore50` is not defined. @ericstj is that right?

Comment (ericstj) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/test-runtime/project.json]: Actually `netstandard1.3` is completely redundant.  NS1.3 is supported by DNXCore50.

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/test-runtime/project.json [hunk 12]
```diff
@@ -11,12 +11,14 @@
   "frameworks": {
     "dnxcore50": {
       "dependencies": {
+        "Microsoft.DotNet.xunit.performance.analysis.cli": "1.0.0-alpha-build0032",
```
Comment (ericstj) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/test-runtime/project.json]: This package depends on 
```
<dependency id="Microsoft.NETCore.App" version="[1.0.0-rc2-3002520, )" />
```
I think you should trim this down to just the packages you actually depend on.  @weshaggard what do you think?

Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/test-runtime/project.json]: Agreed. We don't want our test projects depending on NETCore.App, at least not yet.

**Comment by dsgouda** [comment]
@dotnet-bot Test this please

**Comment by dsgouda** [comment]
@ericstj 
We published a new package for xunit performance runner which takes a dependency on NetStandard libraries. I have referenced it under the runtime setup. PTAL

**Comment by ericstj** [comment]
LGTM
